File: sdks/python/examples/multi_agent_coordination.py
"""
Multi-Agent Coordination Example

Shows how to use OverseeX coordination intelligence features
for monitoring multi-agent systems.
"""

import os
from overseex import OverseeX


def main():
    client = OverseeX(
        api_key=os.environ.get("OVERSEEX_API_KEY", "ox_test_xxx"),
    )

    # Example 1: Record agent handoffs in a multi-agent workflow
    with client.span("orchestrator_agent", tags=["multi-agent", "orchestrator"]) as span:
        span.set_input({"task": "Research and summarize AI news"})

        # Delegate to research agent
        span.record_handoff(
            from_agent="orchestrator-001",
            to_agent="researcher-001",
            context={"task": "Find recent AI news articles"},
            handoff_type="delegation",
        )

        # Simulate research agent work (in real code, this would be async)
        import time
        time.sleep(0.1)

        # Research agent hands back to orchestrator
        span.record_handoff(
            from_agent="researcher-001",
            to_agent="orchestrator-001",
            context={"articles": ["article1", "article2", "article3"]},
            handoff_type="completion",
        )

        # Delegate to summarizer agent
        span.record_handoff(
            from_agent="orchestrator-001",
            to_agent="summarizer-001",
            context={
                "articles": ["article1", "article2", "article3"],
                "format": "bullet_points",
            },
            handoff_type="delegation",
        )

        time.sleep(0.05)

        # Summarizer completes
        span.record_handoff(
            from_agent="summarizer-001",
            to_agent="orchestrator-001",
            context={"summary": "• AI advancement 1\n• AI advancement 2"},
            handoff_type="completion",
        )

        span.set_output({"summary": "• AI advancement 1\n• AI advancement 2"})
        span.set_status("success")

    # Example 2: Check for coordination issues
    print("\n--- Checking Coordination Issues ---")
    issues = client.coordination.list_issues(
        severity="high",
        status="open",
        limit=10,
    )

    if issues:
        print(f"Found {len(issues)} open high-severity issues:")
        for issue in issues:
            print(f"  - [{issue.issue_type}] {issue.description}")
    else:
        print("No high-severity issues found!")

    # Example 3: Get corrective suggestions
    print("\n--- Checking Suggestions ---")
    suggestions = client.coordination.list_suggestions(
        min_confidence=0.7,
        status="pending",
    )

    if suggestions:
        print(f"Found {len(suggestions)} suggestions:")
        for suggestion in suggestions:
            print(
                f"  - [{suggestion.confidence * 100:.0f}%] {suggestion.suggestion_text}"
            )

            # Example: Approve a suggestion
            # client.coordination.approve_suggestion(
            #     suggestion.id,
            #     feedback_notes="Implemented successfully"
            # )
    else:
        print("No pending suggestions found!")

    # Example 4: View learned patterns
    print("\n--- Learned Patterns ---")
    patterns = client.coordination.list_patterns(is_active=True, limit=5)

    if patterns:
        print(f"{len(patterns)} active patterns:")
        for pattern in patterns:
            print(
                f"  - [{pattern.issue_type}] {pattern.pattern_name} "
                f"({pattern.success_rate * 100:.0f}% success)"
            )
    else:
        print("No patterns learned yet.")

    # Example 5: Get coordination metrics
    print("\n--- Coordination Metrics ---")
    metrics = client.coordination.get_metrics(days=7)
    print("Last 7 days:")
    print(f"  Total issues: {metrics.total_issues}")
    print(f"  Issues by type: {metrics.issues_by_type}")
    print(f"  Total handoffs: {metrics.total_handoffs}")
    print(f"  Handoff success rate: {metrics.handoff_success_rate * 100:.1f}%")

    # Example 6: Analyze specific traces for issues
    print("\n--- Analyzing Traces ---")
    # Uncomment to analyze traces:
    # analysis = client.coordination.analyze_traces(
    #     trace_ids=["trace-1", "trace-2"],
    #     auto_create_issues=True,
    # )
    # print(f"Analysis result: {analysis}")

    print("\nCoordination example complete!")


if __name__ == "__main__":
    main()

File: sdks/typescript/examples/multi-agent-coordination.ts
/**
 * Multi-Agent Coordination Example
 *
 * Shows how to use OverseeX coordination intelligence features
 * for monitoring multi-agent systems.
 */

import { OverseeX } from '../src';

async function main() {
  const client = new OverseeX({
    apiKey: process.env.OVERSEEX_API_KEY || 'ox_test_xxx',
  });

  // Example 1: Record agent handoffs
  const orchestratorAgent = client.createSpan('orchestrator_agent', {
    tags: ['multi-agent', 'orchestrator'],
  });

  try {
    orchestratorAgent.setInput({ task: 'Research and summarize AI news' });

    // Delegate to research agent
    orchestratorAgent.recordHandoff(
      'orchestrator-001',
      'researcher-001',
      { task: 'Find recent AI news articles' },
      'delegation'
    );

    // Simulate research agent work
    await new Promise((resolve) => setTimeout(resolve, 100));

    // Research agent hands back to orchestrator
    orchestratorAgent.recordHandoff(
      'researcher-001',
      'orchestrator-001',
      { articles: ['article1', 'article2', 'article3'] },
      'completion'
    );

    // Delegate to summarizer agent
    orchestratorAgent.recordHandoff(
      'orchestrator-001',
      'summarizer-001',
      { articles: ['article1', 'article2', 'article3'], format: 'bullet_points' },
      'delegation'
    );

    // Simulate summarizer work
    await new Promise((resolve) => setTimeout(resolve, 50));

    // Summarizer completes
    orchestratorAgent.recordHandoff(
      'summarizer-001',
      'orchestrator-001',
      { summary: '• AI advancement 1\n• AI advancement 2\n• AI advancement 3' },
      'completion'
    );

    orchestratorAgent.setOutput({
      summary: '• AI advancement 1\n• AI advancement 2\n• AI advancement 3',
    });
    orchestratorAgent.setStatus('success');
  } finally {
    await orchestratorAgent.end(client);
  }

  // Example 2: Check for coordination issues
  console.log('\n--- Checking Coordination Issues ---');
  const issues = await client.coordination.listIssues({
    severity: 'high',
    status: 'open',
    limit: 10,
  });

  if (issues.length > 0) {
    console.log(`Found ${issues.length} open high-severity issues:`);
    for (const issue of issues) {
      console.log(`  - [${issue.issue_type}] ${issue.description}`);
    }
  } else {
    console.log('No high-severity issues found!');
  }

  // Example 3: Get corrective suggestions
  console.log('\n--- Checking Suggestions ---');
  const suggestions = await client.coordination.listSuggestions({
    minConfidence: 0.7,
    status: 'pending',
  });

  if (suggestions.length > 0) {
    console.log(`Found ${suggestions.length} suggestions:`);
    for (const suggestion of suggestions) {
      console.log(
        `  - [${(suggestion.confidence * 100).toFixed(0)}%] ${suggestion.suggestion_text}`
      );

      // Example: Approve a suggestion
      // await client.coordination.approveSuggestion(
      //   suggestion.id,
      //   'Implemented successfully'
      // );
    }
  } else {
    console.log('No pending suggestions found!');
  }

  // Example 4: View learned patterns
  console.log('\n--- Learned Patterns ---');
  const patterns = await client.coordination.listPatterns({
    isActive: true,
    limit: 5,
  });

  if (patterns.length > 0) {
    console.log(`${patterns.length} active patterns:`);
    for (const pattern of patterns) {
      console.log(
        `  - [${pattern.issue_type}] ${pattern.pattern_name} (${pattern.success_rate * 100}% success)`
      );
    }
  } else {
    console.log('No patterns learned yet.');
  }

  // Example 5: Get coordination metrics
  console.log('\n--- Coordination Metrics ---');
  const metrics = await client.coordination.getMetrics({ days: 7 });
  console.log('Last 7 days:');
  console.log(`  Total issues: ${metrics.total_issues}`);
  console.log(`  Issues by type:`, metrics.issues_by_type);
  console.log(`  Total handoffs: ${metrics.total_handoffs}`);
  console.log(`  Handoff success rate: ${(metrics.handoff_success_rate * 100).toFixed(1)}%`);
}

main().catch(console.error);
